File: engines/csharp/src/Engine/Replay.cs
using System;

namespace G2048.Engine;

/// <summary>Replay verification failure (SPEC §8). MoveIndex is -1 when not applicable.</summary>
public sealed class ReplayException(string code, string message, int moveIndex) : Exception(message)
{
    public const string InvalidMoveAt = "INVALID_MOVE_AT";
    public const string BadLetter = "BAD_LETTER";
    public const string SpecVersionError = "SPEC_VERSION";
    public const string FinalMismatch = "FINAL_MISMATCH";

    public string Code { get; } = code;
    public int MoveIndex { get; } = moveIndex;
}

/// <summary>Client-claimed final values; null fields are not checked.</summary>
public sealed record FinalClaim(string? Board = null, long? Score = null, int? MoveCount = null, string? HistoryHash = null);

public static class Replay
{
    public static int LetterToDirection(char ch, int at)
    {
        int d = Board.DirectionLetters.IndexOf(ch);
        if (d < 0) throw new ReplayException(ReplayException.BadLetter, $"bad move letter '{ch}' at {at}", at);
        return d;
    }

    /// <summary>Re-plays moves from seed.</summary>
    public static Game Simulate(uint seed, string moves)
    {
        var g = new Game(seed);
        for (int i = 0; i < moves.Length; i++)
        {
            int d = LetterToDirection(moves[i], i);
            if (!g.Apply(d)) throw new ReplayException(ReplayException.InvalidMoveAt, $"INVALID_MOVE_AT {i}", i);
        }
        return g;
    }

    /// <summary>Re-simulates a replay and returns the authoritative snapshot.</summary>
    public static Snapshot Verify(int specVersion, uint seed, string moves, FinalClaim? final)
    {
        if (specVersion != Game.SpecVersion)
            throw new ReplayException(ReplayException.SpecVersionError, $"unsupported specVersion {specVersion}", -1);
        var snap = Simulate(seed, moves).Snapshot();
        if (final is null) return snap;
        static ReplayException Mismatch(string key, object claimed, object actual) =>
            new(ReplayException.FinalMismatch, $"final.{key} mismatch: claimed {claimed}, actual {actual}", -1);
        if (final.Board is not null && final.Board != snap.Board) throw Mismatch("board", final.Board, snap.Board);
        if (final.Score is long s && s != snap.Score) throw Mismatch("score", s, snap.Score);
        if (final.MoveCount is int m && m != snap.MoveCount) throw Mismatch("moveCount", m, snap.MoveCount);
        if (final.HistoryHash is not null && final.HistoryHash != snap.HistoryHash)
            throw Mismatch("historyHash", final.HistoryHash, snap.HistoryHash);
        return snap;
    }
}

File: engines/go/engine/replay.go
package engine

import (
	"fmt"
	"strings"
)

// Replay error codes (SPEC §8).
const (
	ErrInvalidMoveAt = "INVALID_MOVE_AT"
	ErrBadLetter     = "BAD_LETTER"
	ErrSpecVersion   = "SPEC_VERSION"
	ErrFinalMismatch = "FINAL_MISMATCH"
)

// ReplayError is a replay verification failure. MoveIndex is -1 when not applicable.
type ReplayError struct {
	Code      string
	Message   string
	MoveIndex int
}

func (e *ReplayError) Error() string { return e.Message }

// LetterToDirection maps a replay letter to a direction.
func LetterToDirection(ch byte, at int) (int, error) {
	d := strings.IndexByte(DirectionLetters, ch)
	if d < 0 {
		return -1, &ReplayError{Code: ErrBadLetter, Message: fmt.Sprintf("bad move letter '%c' at %d", ch, at), MoveIndex: at}
	}
	return d, nil
}

// Simulate re-plays moves from seed.
func Simulate(seed uint32, moves string) (*Game, error) {
	g := NewGame(seed)
	for i := 0; i < len(moves); i++ {
		d, err := LetterToDirection(moves[i], i)
		if err != nil {
			return nil, err
		}
		if _, ok := g.Apply(d); !ok {
			return nil, &ReplayError{Code: ErrInvalidMoveAt, Message: fmt.Sprintf("INVALID_MOVE_AT %d", i), MoveIndex: i}
		}
	}
	return g, nil
}

// FinalClaim holds client-claimed final values; nil fields are not checked.
type FinalClaim struct {
	Board       *string `json:"board,omitempty"`
	Score       *int64  `json:"score,omitempty"`
	MoveCount   *int    `json:"moveCount,omitempty"`
	HistoryHash *string `json:"historyHash,omitempty"`
}

// VerifyReplay re-simulates a replay and returns the authoritative snapshot.
func VerifyReplay(specVersion int, seed uint32, moves string, final *FinalClaim) (Snapshot, error) {
	if specVersion != SpecVersion {
		return Snapshot{}, &ReplayError{Code: ErrSpecVersion, Message: fmt.Sprintf("unsupported specVersion %d", specVersion), MoveIndex: -1}
	}
	g, err := Simulate(seed, moves)
	if err != nil {
		return Snapshot{}, err
	}
	snap := g.Snapshot()
	if final != nil {
		mismatch := func(key string, claimed, actual any) error {
			return &ReplayError{Code: ErrFinalMismatch, Message: fmt.Sprintf("final.%s mismatch: claimed %v, actual %v", key, claimed, actual), MoveIndex: -1}
		}
		if final.Board != nil && *final.Board != snap.Board {
			return snap, mismatch("board", *final.Board, snap.Board)
		}
		if final.Score != nil && *final.Score != snap.Score {
			return snap, mismatch("score", *final.Score, snap.Score)
		}
		if final.MoveCount != nil && *final.MoveCount != snap.MoveCount {
			return snap, mismatch("moveCount", *final.MoveCount, snap.MoveCount)
		}
		if final.HistoryHash != nil && *final.HistoryHash != snap.HistoryHash {
			return snap, mismatch("historyHash", *final.HistoryHash, snap.HistoryHash)
		}
	}
	return snap, nil
}
